### PR TITLE
autotools: extend patching of the libtool script

### PR DESCRIPTION
**Pull request by skosukhin**
This introduces the argument `start_at` for the `filter_file` function. The value of the argument is used as a marker to start applying the replacements. If a text line matches this marker filtering is started at the next line. All contents before the marker and the marker itself are copied verbatim. The default is to start filtering from the first line of the file.

This feature allows for more selective filtering of the libtool script. For example, currently, the empty PIC flag entries in a libtool script are filled with the flag for the C compiler (`cc_pic_flag`). The suggested modification fills the values with the language-specific flags (see NAG).

I have also extended the method `_do_patch_libtool` with fixes that currently live separately in several packages (e.g. `mpich`, `openmpi`, `netcdf-fortran`, `parallel-netcdf`, etc.). I plan to go over those packages later and remove the patches there but it looks like I have managed to implement the filtering in a way it can co-exist with similar patches that either are applied with Spack or come with the packages (e.g. `openmpi`, `pmix`).

### CODE REVIEW

**Comment by skosukhin** [comment]
And here is an example of what can be finally removed: #30769.

**Comment by alalazo** [comment]
@skosukhin I was too hasty in merging #30769 I'd be willing to merge this if it can treat those kind of patches automatically, and the `start_at` argument LGTM. The only thing I'd need before I can merge this is @scottwittenburg to tell me how to trigger a complete rebuild of everything in pipelines, to check software builds are not breaking

**Comment by alalazo** [comment]
If that takes too much time I'll revert #30769 and resubmit it again a s a PR Let me know.

**Comment by skosukhin** [comment]
@alalazo kind reminder

**Comment by alalazo** [comment]
@scottwittenburg Can we do a complete rebuild of the software in the pipelines for this PR? This change seems good in general, but it affects all autotools projects. I want to be sure that all pipelines are building fine from scratch before merging it.

File: lib/spack/spack/build_systems/autotools.py
Comment (alalazo) [inline lib/spack/spack/build_systems/autotools.py]: For easier reading can we add keywords here:
```suggestion
        x.filter(regex=r'^(\s*if test x-L = )("\$p" \|\|\s*)$', repl=r'\1x\2')
        x.filter(regex=r'^(\s*test x-R = )("\$p")(; then\s*)$',
                 repl=r'\1x\2 || test x-l = x"$p"\3')
        # Support Libtool 2.4.2 and older:
        x.filter(regex=r'^(\s*test \$p = "-R")(; then\s*)$',
                 repl=r'\1 || test x-l = x"$p"\2')
```
and below?

**Comment by skosukhin** [comment]
@alalazo I have rebased the branch and addressed your request for changes. Just a reminder that we still cannot build MPICH with NAG without this :(

**Comment by skosukhin** [comment]
@haampie I assume that I am asked to do the same thing here that you did in #32086. Could you, please, help me and check whether I'm doing it right because I do not fully understand what 6ef0643 is supposed to do? Should I tell the bot to run the pipeline? What does that pipeline check normally and how exactly the aforementioned commit alters the tests? Thank you!

**Comment by scottwittenburg** [comment]
@skosukhin Looks what you did will work to force everything to rebuild, sorry it's a pain, we'll be able to do this easier soon.  Note that we don't run the gitlab pipeline until your PR passes the style check (so you're all set there), and until it is ready for review (i.e. no longer a draft).

**Comment by skosukhin** [comment]
So, I now need help in understanding the results of the tests. Some of them failed but I don't see how that can be connected to my changes. Should I simply rebase?

**Comment by alalazo** [comment]
@skosukhin I have seen a few packages where [`makeinfo` is missing](https://gitlab.spack.io/spack/spack/-/jobs/3074075)? In that case I'd try to solve the issues one-by-one in separate PRs after having found a solution following the instructions at the bottom:
```
To reproduce this build locally, run:
    spack ci reproduce-build https://gitlab.spack.io/api/v4/projects/2/jobs/3074075/artifacts [--working-dir <dir>]
If this project does not have public pipelines, you will need to first:
    export GITLAB_PRIVATE_TOKEN=<generated_token>
... then follow the printed instructions.
```
Let me know if you need help with that, I'll try to help pushing this PR forward. Once the issues we see are solved in `develop` I'd rebase this one and see what happens.

**Comment by skosukhin** [comment]
@alalazo this actually might be related to my modifications, which change the timestamps with the patching and trigger the rules that require `makeinfo`. I'll check.

**Comment by skosukhin** [comment]
My assumption was correct. I thought it was not the case but there might be distributed automatically generated files that depend on the `configure` script. In the case of `autoconf-archive`, it's `version.texi`. It's funny that the recipe that generates the file is written in a way that if the contents of the generated file are the same as the contents of the existing one, the existing one does not get updated. However, the contents of the file do change if we simply touch the configure script:
```patch
-@set UPDATED 11 February 2022
+@set UPDATED 10 February 2022
```
and that triggers rules that need `makeinfo` (package `texinfo`, if I'm not mistaken).

So, there is a way to cure the symptom, which I added in #32232 (it also speeds up the installation by omitting the unnecessary patching of the configure script). And there is a solution to use [`keep_modification_time`](https://github.com/spack/spack/blob/7f81a0a25ef032aafb132d499d469fdaeaeb3ebe/lib/spack/llnl/util/filesystem.py#L2307) when patching the configure script, which I'm going to implement here soon.

**Comment by skosukhin** [comment]
That's an interesting game. Turns out, we should not try to patch directories that happen to have names `libtool` and `configure`.

**Comment by alalazo** [comment]
@skosukhin Is the `py-numpy` failure also related to the PR?

**Comment by skosukhin** [comment]
@alalazo I don't know, sorry, haven't had time yet to check.

**Comment by skosukhin** [comment]
@alalazo I have a hard time understanding the reason for [this](https://gitlab.spack.io/spack/spack/-/jobs/3076395):
```
==> [2022-08-18-15:56:45.647844] Fetching s3://spack-binaries-prs/pr30768_libtool_patching/build_cache/linux-amzn2-graviton2/gcc-7.3.1/py-cython-0.29.30/linux-amzn2-graviton2-gcc-7.3.1-py-cython-0.29.30-uhuhx6mmeznpf6w64gky772x66byc5ze.spack
==> [2022-08-18-15:56:45.864869] Flagging py-cython-0.29.30-uhuhx6mmeznpf6w64gky772x66byc5ze as failed: Package tarball failed checksum verification.
It cannot be installed.
==> [2022-08-18-15:56:45.873063] Warning: Skipping build of py-numpy-1.23.2-zrqv3xp7bsiyniytkv5jwi3c7kybbytr since py-cython-0.29.30-uhuhx6mmeznpf6w64gky772x66byc5ze failed
==> [2022-08-18-15:56:45.873110] Flagging py-numpy-1.23.2-zrqv3xp7bsiyniytkv5jwi3c7kybbytr as failed
==> [2022-08-18-15:56:45.881717] Removing build task for py-numpy-1.23.2-zrqv3xp7bsiyniytkv5jwi3c7kybbytr from list
==> [2022-08-18-15:56:45.881801] Error: Failed to install py-cython due to NoChecksumException: Package tarball failed checksum verification.
It cannot be installed.
```
Why is it trying to use the binary distribution? Isn't it supposed to build everything?

**Comment by scottwittenburg** [comment]
> Why is it trying to use the binary distribution? Isn't it supposed to build everything?

Even when building everything, any particular job in the pipeline still needs to install its dependencies from binary.  However, this error has nothing to do with your PR and is caused by another problem we hope to have fixed this week.

**Comment by skosukhin** [comment]
@alalazo do you think the tests are green enough now? The coverage test does not like this PR after the rebase.

**Comment by skosukhin** [comment]
@alalazo shall we get this merged? :)

**UPD** Oh, I have to resolve conflicts first. And re-run the tests again :(

**Comment by haampie** [comment]
@spackbot run pipeline

unsurprisingly gitlab pipelines have false negatives :(

**Comment by haampie** [comment]
~I'm afraid~ this is ~not~ entirely correct for `libtool --tag=CC`. There's no `BEGIN LIBTOOL TAG CONFIG: CC` because it's in the default libtool config.

The generated `libtool` file in pseudo-shell script is pretty much:

```bash
#!/bin/bash

# ### BEGIN LIBTOOL CONFIG
my_variable="my CC value"
# ### END LIBTOOL CONFIG

if --tag=CXX; then
    eval [the CXX block below, retrieved with sed]
elif --tag=FC; then
    eval [the FC block below ...]
fi

do_stuff

exit 0

# notice that exit prevents the stuff below from being executed

# ### BEGIN LIBTOOL TAG CONFIG: CXX
my_variable="CXX value"
# ### END LIBTOOL TAG CONFIG: CXX

# ### BEGIN LIBTOOL TAG CONFIG: FC
my_variable="FC value"
# ### END LIBTOOL TAG CONFIG: FC
```

So for `CC` you shouldn't be looking for the tag, but rather change stuff in the `BEGIN LIBTOOL CONFIG` bit.

Edit: sorry, I read the PR the wrong way -- you're doing this already :sweat_smile:

**Comment by haampie** [comment]
Thanks for this PR, it's a big improvement over replacing strings without context in the libtool file.

1. Can you drop the last gitlab pipeline commit?
2. Can you document `_do_patch_libtool` for posterity? Something along the lines of: 
   > `libtool` supports mixed compilers through tags: `libtool --tag=CC/CXX/FC/...`. For each tag there is a block with variables, which defines what flags to pass to the compiler. The default variables (which are used with `--tag=CC`) are set in a block enclosed by `# ### {BEGIN,END} LIBTOOL CONFIG`. For other tags, there are corresponding blocks `# ### {BEGIN,END} LIBTOOL TAG CONFIG: {CXX,FC,F77}` and the end of the file (after the `exit` command). `libtool` evals those blocks. Whenever we need to update variables that the configure script got wrong (for example cause it did not recognize the compiler), we should properly scope those changes to these tags/blocks.

**Review by haampie — CHANGES_REQUESTED** [review]
Thanks for this PR, it's a big improvement over replacing strings without context in the libtool file.

1. Can you drop the last gitlab pipeline commit?
2. Can you document `_do_patch_libtool` for posterity? Something along the lines of: 
   > `libtool` supports mixed compilers through tags: `libtool --tag=CC/CXX/FC/...`. For each tag there is a block with variables, which defines what flags to pass to the compiler. The default variables (which are used with `--tag=CC`) are set in a block enclosed by `# ### {BEGIN,END} LIBTOOL CONFIG`. For other tags, there are corresponding blocks `# ### {BEGIN,END} LIBTOOL TAG CONFIG: {CXX,FC,F77}` at the end of the file (after the `exit` command). `libtool` evals those blocks based on what --tag is passed. Whenever we need to update variables that the configure script got wrong (for example cause it did not recognize the compiler), we should properly scope those changes to these tags/blocks.

**Review by haampie — APPROVED** [review]

**Pull request merged**
